File: backend/app/services/document_service.py
import os
import uuid
import aiofiles
from typing import List, Optional, Dict, Any
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, and_, func, desc
from sqlalchemy.orm import selectinload
from fastapi import HTTPException, UploadFile, status
from datetime import datetime

from app.models.document import Document, DocumentType, DocumentStatus
from app.models.student import Student
from app.models.user import User
from app.schemas.document import DocumentCreate, DocumentUpdate, DocumentVerification
from app.core.config import settings


class DocumentService:
    """Service for managing student documents"""
    
    ALLOWED_EXTENSIONS = {
        '.pdf', '.doc', '.docx', '.jpg', '.jpeg', '.png', '.gif', '.webp', '.txt'
    }
    MAX_FILE_SIZE = 10 * 1024 * 1024  # 10MB
    
    @staticmethod
    def _get_file_extension(filename: str) -> str:
        """Get file extension from filename"""
        return os.path.splitext(filename)[1].lower()
    
    @staticmethod
    def _validate_file(file: UploadFile) -> None:
        """Validate uploaded file"""
        if not file.filename:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="No file provided"
            )
        
        # Check file extension
        ext = DocumentService._get_file_extension(file.filename)
        if ext not in DocumentService.ALLOWED_EXTENSIONS:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"File type {ext} not allowed. Allowed types: {', '.join(DocumentService.ALLOWED_EXTENSIONS)}"
            )
        
        # Check file size
        if file.size and file.size > DocumentService.MAX_FILE_SIZE:
            raise HTTPException(
                status_code=status.HTTP_413_REQUEST_ENTITY_TOO_LARGE,
                detail=f"File size too large. Maximum size is {DocumentService.MAX_FILE_SIZE // (1024*1024)}MB"
            )
    
    @staticmethod
    async def upload_document(
        db: AsyncSession,
        file: UploadFile,
        document_data: DocumentCreate,
        uploaded_by: str,
        school_id: str
    ) -> Document:
        """Upload a new document"""
        # Validate file
        DocumentService._validate_file(file)
        
        # Verify student exists and belongs to school
        result = await db.execute(
            select(Student).where(
                and_(
                    Student.id == document_data.student_id,
                    Student.school_id == school_id,
                    Student.is_deleted == False
                )
            )
        )
        student = result.scalar_one_or_none()
        if not student:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Student not found"
            )
        
        # Create upload directory
        upload_dir = os.path.join(settings.upload_dir, "documents", school_id)
        os.makedirs(upload_dir, exist_ok=True)
        
        # Generate unique filename
        ext = DocumentService._get_file_extension(file.filename)
        unique_filename = f"{uuid.uuid4().hex}{ext}"
        file_path = os.path.join(upload_dir, unique_filename)
        
        try:
            # Save file
            content = await file.read()
            async with aiofiles.open(file_path, 'wb') as f:
                await f.write(content)
            
            # Create document record
            document = Document(
                title=document_data.title,
                description=document_data.description,
                document_type=document_data.document_type,
                file_name=unique_filename,
                original_file_name=file.filename,
                file_path=file_path,
                file_size=len(content),
                mime_type=file.content_type or 'application/octet-stream',
                uploaded_by=uploaded_by,
                student_id=document_data.student_id,
                expires_at=document_data.expires_at,
                tags=document_data.tags,
                is_public=document_data.is_public,
                school_id=school_id
            )
            
            db.add(document)
            await db.commit()
            await db.refresh(document)
            
            return document
            
        except Exception as e:
            # Clean up file if it was created
            if os.path.exists(file_path):
                os.remove(file_path)
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail=f"Failed to upload document: {str(e)}"
            )
    
    @staticmethod
    async def get_documents(
        db: AsyncSession,
        school_id: str,
        student_id: Optional[str] = None,
        document_type: Optional[DocumentType] = None,
        status: Optional[DocumentStatus] = None,
        skip: int = 0,
        limit: int = 100
    ) -> List[Document]:
        """Get documents with filtering"""
        query = select(Document).options(
            selectinload(Document.uploader),
            selectinload(Document.verifier),
            selectinload(Document.student)
        ).where(
            and_(
                Document.school_id == school_id,
                Document.is_deleted == False
            )
        )
        
        if student_id:
            query = query.where(Document.student_id == student_id)
        if document_type:
            query = query.where(Document.document_type == document_type)
        if status:
            query = query.where(Document.status == status)
        
        query = query.order_by(desc(Document.created_at)).offset(skip).limit(limit)
        
        result = await db.execute(query)
        return result.scalars().all()
    
    @staticmethod
    async def get_document_by_id(
        db: AsyncSession,
        document_id: str,
        school_id: str
    ) -> Optional[Document]:
        """Get document by ID"""
        result = await db.execute(
            select(Document).options(
                selectinload(Document.uploader),
                selectinload(Document.verifier),
                selectinload(Document.student)
            ).where(
                and_(
                    Document.id == document_id,
                    Document.school_id == school_id,
                    Document.is_deleted == False
                )
            )
        )
        return result.scalar_one_or_none()
    
    @staticmethod
    async def update_document(
        db: AsyncSession,
        document_id: str,
        document_data: DocumentUpdate,
        school_id: str
    ) -> Optional[Document]:
        """Update document metadata"""
        document = await DocumentService.get_document_by_id(db, document_id, school_id)
        if not document:
            return None
        
        # Update fields
        for field, value in document_data.dict(exclude_unset=True).items():
            setattr(document, field, value)
        
        document.updated_at = datetime.utcnow()
        await db.commit()
        await db.refresh(document)
        
        return document
    
    @staticmethod
    async def verify_document(
        db: AsyncSession,
        document_id: str,
        verification_data: DocumentVerification,
        verified_by: str,
        school_id: str
    ) -> Optional[Document]:
        """Verify or reject a document"""
        document = await DocumentService.get_document_by_id(db, document_id, school_id)
        if not document:
            return None
        
        document.status = verification_data.status
        document.verification_notes = verification_data.verification_notes
        document.verified_by = verified_by
        document.verified_at = datetime.utcnow()
        document.updated_at = datetime.utcnow()
        
        await db.commit()
        await db.refresh(document)
        
        return document
    
    @staticmethod
    async def delete_document(
        db: AsyncSession,
        document_id: str,
        school_id: str
    ) -> bool:
        """Soft delete document"""
        document = await DocumentService.get_document_by_id(db, document_id, school_id)
        if not document:
            return False
        
        # Soft delete
        document.is_deleted = True
        document.updated_at = datetime.utcnow()
        
        # Optionally delete physical file
        try:
            if os.path.exists(document.file_path):
                os.remove(document.file_path)
        except Exception:
            pass  # Don't fail if file deletion fails
        
        await db.commit()
        return True
    
    @staticmethod
    async def get_document_stats(
        db: AsyncSession,
        school_id: str,
        student_id: Optional[str] = None
    ) -> Dict[str, Any]:
        """Get document statistics"""
        base_query = select(Document).where(
            and_(
                Document.school_id == school_id,
                Document.is_deleted == False
            )
        )
        
        if student_id:
            base_query = base_query.where(Document.student_id == student_id)
        
        # Total documents
        total_result = await db.execute(
            select(func.count(Document.id)).select_from(base_query.subquery())
        )
        total_documents = total_result.scalar() or 0
        
        # Documents by status
        status_result = await db.execute(
            select(Document.status, func.count(Document.id))
            .select_from(base_query.subquery())
            .group_by(Document.status)
        )
        status_counts = dict(status_result.all())
        
        # Documents by type
        type_result = await db.execute(
            select(Document.document_type, func.count(Document.id))
            .select_from(base_query.subquery())
            .group_by(Document.document_type)
        )
        type_counts = dict(type_result.all())
        
        # Recent uploads
        recent_result = await db.execute(
            base_query.options(
                selectinload(Document.uploader),
                selectinload(Document.student)
            ).order_by(desc(Document.created_at)).limit(5)
        )
        recent_documents = recent_result.scalars().all()
        
        return {
            'total_documents': total_documents,
            'pending_documents': status_counts.get(DocumentStatus.PENDING, 0),
            'approved_documents': status_counts.get(DocumentStatus.APPROVED, 0),
            'rejected_documents': status_counts.get(DocumentStatus.REJECTED, 0),
            'expired_documents': status_counts.get(DocumentStatus.EXPIRED, 0),
            'documents_by_type': type_counts,
            'recent_uploads': recent_documents
        }

File: frontend/src/services/documentService.ts
import { apiService } from './api';

export interface Document {
  id: string;
  title: string;
  description?: string;
  document_type: DocumentType;
  status: DocumentStatus;
  file_name: string;
  original_file_name: string;
  file_size: number;
  mime_type: string;
  uploaded_by: string;
  student_id: string;
  verified_by?: string;
  verified_at?: string;
  verification_notes?: string;
  expires_at?: string;
  tags?: string;
  is_public: boolean;
  school_id: string;
  created_at: string;
  updated_at: string;
  file_size_mb?: number;
  is_expired?: boolean;
  is_image?: boolean;
  is_pdf?: boolean;
  uploader_name?: string;
  verifier_name?: string;
  student_name?: string;
}

export type DocumentType = 
  | 'birth_certificate'
  | 'passport'
  | 'national_id'
  | 'medical_record'
  | 'academic_transcript'
  | 'immunization_record'
  | 'photo'
  | 'parent_id'
  | 'proof_of_address'
  | 'other';

export type DocumentStatus = 'pending' | 'approved' | 'rejected' | 'expired';

export interface DocumentUploadData {
  title: string;
  student_id: string;
  document_type: DocumentType;
  description?: string;
  expires_at?: string;
  tags?: string;
  is_public?: boolean;
}

export interface DocumentUpdateData {
  title?: string;
  description?: string;
  document_type?: DocumentType;
  expires_at?: string;
  tags?: string;
  is_public?: boolean;
  status?: DocumentStatus;
  verification_notes?: string;
}

export interface DocumentVerificationData {
  status: DocumentStatus;
  verification_notes?: string;
}

export interface DocumentUploadResponse {
  message: string;
  document_id: string;
  file_url: string;
}

export interface DocumentStats {
  total_documents: number;
  pending_documents: number;
  approved_documents: number;
  rejected_documents: number;
  expired_documents: number;
  documents_by_type: Record<string, number>;
  recent_uploads: Document[];
}

class DocumentService {
  // Upload a document
  async uploadDocument(file: File, data: DocumentUploadData): Promise<DocumentUploadResponse> {
    const formData = new FormData();
    formData.append('file', file);
    formData.append('title', data.title);
    formData.append('student_id', data.student_id);
    formData.append('document_type', data.document_type);
    
    if (data.description) formData.append('description', data.description);
    if (data.expires_at) formData.append('expires_at', data.expires_at);
    if (data.tags) formData.append('tags', data.tags);
    if (data.is_public !== undefined) formData.append('is_public', data.is_public.toString());

    return apiService.post<DocumentUploadResponse>('/api/v1/documents/upload', formData, {
      headers: {
        'Content-Type': 'multipart/form-data',
      },
    });
  }

  // Get documents with filtering
  async getDocuments(params?: {
    student_id?: string;
    document_type?: DocumentType;
    status?: DocumentStatus;
    skip?: number;
    limit?: number;
  }): Promise<Document[]> {
    const queryParams = new URLSearchParams();
    if (params?.student_id) queryParams.append('student_id', params.student_id);
    if (params?.document_type) queryParams.append('document_type', params.document_type);
    if (params?.status) queryParams.append('status', params.status);
    if (params?.skip) queryParams.append('skip', params.skip.toString());
    if (params?.limit) queryParams.append('limit', params.limit.toString());

    const url = `/api/v1/documents/${queryParams.toString() ? '?' + queryParams.toString() : ''}`;
    return apiService.get<Document[]>(url);
  }

  // Get document by ID
  async getDocumentById(documentId: string): Promise<Document> {
    return apiService.get<Document>(`/api/v1/documents/${documentId}`);
  }

  // Get student documents
  async getStudentDocuments(studentId: string, params?: {
    document_type?: DocumentType;
    status?: DocumentStatus;
  }): Promise<Document[]> {
    const queryParams = new URLSearchParams();
    if (params?.document_type) queryParams.append('document_type', params.document_type);
    if (params?.status) queryParams.append('status', params.status);

    const url = `/api/v1/documents/student/${studentId}${queryParams.toString() ? '?' + queryParams.toString() : ''}`;
    return apiService.get<Document[]>(url);
  }

  // Update document
  async updateDocument(documentId: string, data: DocumentUpdateData): Promise<Document> {
    return apiService.put<Document>(`/api/v1/documents/${documentId}`, data);
  }

  // Verify document
  async verifyDocument(documentId: string, data: DocumentVerificationData): Promise<Document> {
    return apiService.post<Document>(`/api/v1/documents/${documentId}/verify`, data);
  }

  // Delete document
  async deleteDocument(documentId: string): Promise<void> {
    return apiService.delete(`/api/v1/documents/${documentId}`);
  }



  // Download document
  async downloadDocument(documentId: string): Promise<Blob> {
    const response = await apiService.api.get(`/api/v1/documents/${documentId}/download`, {
      responseType: 'blob',
    });
    return response.data;
  }

  // Get document statistics
  async getDocumentStats(studentId?: string): Promise<DocumentStats> {
    const queryParams = new URLSearchParams();
    if (studentId) queryParams.append('student_id', studentId);

    const url = `/api/v1/documents/stats/overview${queryParams.toString() ? '?' + queryParams.toString() : ''}`;
    return apiService.get<DocumentStats>(url);
  }

  // Get document type display name
  getDocumentTypeDisplayName(type: DocumentType): string {
    const typeNames: Record<DocumentType, string> = {
      birth_certificate: 'Birth Certificate',
      passport: 'Passport',
      national_id: 'National ID',
      medical_record: 'Medical Record',
      academic_transcript: 'Academic Transcript',
      immunization_record: 'Immunization Record',
      photo: 'Photo',
      parent_id: 'Parent ID',
      proof_of_address: 'Proof of Address',
      other: 'Other'
    };
    return typeNames[type] || type;
  }

  // Get status display name and color
  getStatusInfo(status: DocumentStatus): { name: string; color: string } {
    const statusInfo: Record<DocumentStatus, { name: string; color: string }> = {
      pending: { name: 'Pending', color: 'yellow' },
      approved: { name: 'Approved', color: 'green' },
      rejected: { name: 'Rejected', color: 'red' },
      expired: { name: 'Expired', color: 'gray' }
    };
    return statusInfo[status] || { name: status, color: 'gray' };
  }

  // Validate file for upload
  validateFile(file: File): { valid: boolean; error?: string } {
    const maxSize = 10 * 1024 * 1024; // 10MB
    const allowedTypes = [
      'application/pdf',
      'application/msword',
      'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
      'image/jpeg',
      'image/png',
      'image/gif',
      'image/webp',
      'text/plain'
    ];

    if (file.size > maxSize) {
      return { valid: false, error: 'File size must be less than 10MB' };
    }

    if (!allowedTypes.includes(file.type)) {
      return { valid: false, error: 'File type not supported. Please upload PDF, DOC, DOCX, JPG, PNG, GIF, WEBP, or TXT files.' };
    }

    return { valid: true };
  }
}

export const documentService = new DocumentService();
